sdks/go/pkg/beam/core/runtime/harness/harness.go

```diff
@@ -133,7 +129,17 @@ func Main(ctx context.Context, loggingEndpoint, controlEndpoint string) error {
 			if resp != nil {
 				respc <- resp
 			}
-		}()
+		}
+
+		if req.GetProcessBundle() != nil {
+			// Only process bundles in a gorutine. We need to process plan creating etc
```

Comment:
<!--thread_id:cc_166105329_t; commit:977a4f133a10d957b90566426581193d93041321; resolved:0-->
<!--section:context-quote-->
> **wcn3** wrote:
> I'm curious what this need here is. It's certainly simpler to linearize plan creation, but I thought we could handle two plan creations simultaneously, only needing to synchronize on the shared map that holds the plans.

<!--section:body-->
The problem with the old code is that process_bundle and progress_report request instructions were processed while we were still processing the process_bundle_descriptor instruction to create the plan they referred to. We failed bundles with "plan not found" due to this race -- the Dataflow runner doesn't actually wait for a response before sending dependent instructions. You are correct in that we can parallelize everything across plans.

Clarified the comment.